server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/TraceServiceElasticsearch.java

```diff
@@ -314,4 +347,84 @@ public void clear(String tenantId) {
         }
     }
 
-}
+    public static class TraceSerializer extends JsonSerializer<Trace> {
+
+        /* (non-Javadoc)
+         * @see com.fasterxml.jackson.databind.JsonSerializer#serialize(java.lang.Object,
+         *          com.fasterxml.jackson.core.JsonGenerator, com.fasterxml.jackson.databind.SerializerProvider)
+         */
+        @Override
+        public void serialize(Trace trace, JsonGenerator jgen, SerializerProvider provider)
+                throws IOException, JsonProcessingException {
+            jgen.writeStartObject();
+            jgen.writeStringField(BUSINESS_TRANSACTION_FIELD, trace.getBusinessTransaction());
+            jgen.writeStringField(HOST_ADDRESS_FIELD, trace.getHostAddress());
+            jgen.writeStringField(HOST_NAME_FIELD, trace.getHostName());
+            jgen.writeStringField(ID_FIELD, trace.getId());
+            jgen.writeStringField(PRINCIPAL_FIELD, trace.getPrincipal());
+            jgen.writeNumberField(START_TIME_FIELD, trace.getStartTime());
+            jgen.writeArrayFieldStart(NODES_FIELD);
+            for (Node n : trace.getNodes()) {
+                jgen.writeObject(n);
+            }
+            jgen.writeEndArray();
+            Set<Property> properties = trace.allProperties();
+            jgen.writeArrayFieldStart(PROPERTIES_FIELD);
+            for (Property p : properties) {
+                jgen.writeObject(p);
+            }
+            jgen.writeEndArray();
+            jgen.writeEndObject();
+        }
+    }
+
+    public static class TraceDeserializer extends JsonDeserializer<Trace> {
+
+        /* (non-Javadoc)
+         * @see com.fasterxml.jackson.databind.JsonDeserializer#deserialize(com.fasterxml.jackson.core.JsonParser, com.fasterxml.jackson.databind.DeserializationContext)
+         */
+        @Override
+        public Trace deserialize(JsonParser parser, DeserializationContext context)
```

Comment:
As mentioned in the reviewers notes, the reason for explicitly using custom serializer/deserializer is to modify the json stored in Elasticsearch, to aggregate the properties associated with the node hierarchy and associate them with the top level Trace. This is to enable them to be queried, as Elasticsearch doesn't support querying fields on a recursive structure.